src/api/web3.ts

```diff
@@ -0,0 +1,210 @@
+import Web3 from 'web3'
+import { Block } from 'web3-eth'
+import { provider } from 'web3-providers'
+import { TransactionReceipt, Transaction } from 'web3-core'
+
+import { windowLoaded } from './utils'
+
+declare global {
+    interface Window {
+        ethereum: provider & {
+            enable: () => Promise<string[]>
+            on?: () => EventListener
+        }
+        web3: Web3
+    }
+}
+
+export interface Web3FrontendAPI {
+    web3: Web3
+    getCurrentAccount: () => Promise<string>
+    getAccounts: () => Promise<string[]>
+    getBalance: (account?: string) => Promise<string>
+    getBlock: (blockNumberOrString: string | number) => Promise<Partial<Block>>
+    getNetwork: () => Promise<string | number>
+    getNetworkName: (id?: number | string) => Promise<string>
+    getNetworkProviderURL: (id: number | string) => string
+    getTransaction: (txHash: string) => Promise<Transaction>
+    getTransactionReceipt: (txHash: string) => Promise<TransactionReceipt>
+    // Utils
+    isAddress: (address: string) => boolean
+    isReadOnly: () => boolean
+    // Provider Specific
+    resetProvider(): void
+    setProvider(provider: string): void
+}
+
+export interface Web3SetupOptions {
+    customProvider?: provider | string
+    reloadOnAccountsChange?: boolean
+}
+
+let web3API: Web3FrontendAPI
+
+let readOnly: boolean
+
+export const getWeb3API = async (options: Web3SetupOptions = {}, force?: boolean): Promise<Web3FrontendAPI> => {
+    if (!force && web3API) return web3API
+
+    return setupWeb3(options)
+}
+
+const getProvider = async ({ customProvider, reloadOnAccountsChange }: Web3SetupOptions): Promise<Web3> => {
+    let provider: Web3
+
+    // wait window load then check window globals...
+    await windowLoaded
+
+    // Use Custom Provider if passed in
+    if (customProvider) provider = new Web3(customProvider)
```

Comment:
Can we use braces in the ifs? I think it's prone to errors not to add them

We could add this rule: https://eslint.org/docs/rules/curly

Reply:
I don't agree, I think its much more succinct and shorter

Reply:
Compromise:
```
customProvider && provider = new Web3(customProvider)
```
What? You both hate it? No surprise

Serious answer: no real need to add more braces,unless it's in `else` clause

Reply:
Also move it above `await windowLoaded`, no reason to wait if `customProvider`

Reply:
It's consider a bad practice, the linters include it as a rule, most people agree that is simpler to read (Single Line Responsibility principle), it better for the debugger, and it lead to less errors.

It's not a big deal in this example, cause the then clause is short, but I would follow the rule of *always* adding brackets for the given reasons, and because it doesn't cost money to add some intros and bracket